### PR TITLE
[R] Move all DMatrix fields to function arguments

### PR DESCRIPTION
**Pull request by david-cortes**
ref https://github.com/dmlc/xgboost/issues/9810

This PR switches the DMatrix constructor to have all the possible inputs that it accepts as parameters in the function, like the python class for DMatrix does. The docs were all copied from the python DMatrix class.

I've purposefully left out all the parameters related to categorical features as that's still a work in progress from what I understand.

### CODE REVIEW

**Comment by mayer79** [comment]
Fantastic! No doubts anymore if its "weight" or "weights" etc.!

File: R-package/R/xgb.DMatrix.R [hunk 1]
```diff
@@ -34,8 +45,24 @@
...
-xgb.DMatrix <- function(data, info = list(), missing = NA, silent = FALSE, nthread = NULL, ...) {
-  cnames <- NULL
+xgb.DMatrix <- function(
+  data,
+  label = NULL,
+  weight = NULL,
+  base_margin = NULL,
+  missing = NA,
+  silent = FALSE,
+  feature_names = colnames(data),
```
Comment (trivialfis) [inline R-package/R/xgb.DMatrix.R:55]: Nice, didn't know R supports this type of self-referencing parameter.

File: R-package/R/xgb.DMatrix.R [hunk 2]
```diff
@@ -236,24 +287,37 @@ getinfo <- function(object, ...) UseMethod("getinfo")
...
-    stop(
-      "getinfo: name must be one of the following\n",
-      "    'label', 'weight', 'base_margin', 'nrow', 'label_lower_bound', 'label_upper_bound', 'feature_type', 'feature_name'"
-    )
+        !name %in% allowed_fields) {
+    stop("getinfo: name must be one of the following\n",
+         paste(paste0("'", allowed_fields, "'"), collapse = ", "))
   }
-  if (name == "feature_name" || name == "feature_type") {
+  if (name == "nrow") {
```
Comment (trivialfis) [inline R-package/R/xgb.DMatrix.R:304]: Is this useful in practice? We have `dim.xgb.DMatrix` exported.

**Review by trivialfis — COMMENTED** [review]
Overall looks good! One question regarding the shape of the DMatrix.

File: R-package/R/xgb.DMatrix.R (hunk 2 above)
Comment (david-cortes) [reply R-package/R/xgb.DMatrix.R:304]: Then I guess it's not really a must-have.

**Review by trivialfis — APPROVED** [review]

File: R-package/R/xgb.DMatrix.R [hunk 3]
```diff
@@ -44,7 +71,6 @@ xgb.DMatrix <- function(data, info = list(), missing = NA, silent = FALSE, nthre
     handle <- .Call(XGDMatrixCreateFromFile_R, data, as.integer(silent))
   } else if (is.matrix(data)) {
     handle <- .Call(XGDMatrixCreateFromMat_R, data, missing, as.integer(NVL(nthread, -1)))
-    cnames <- colnames(data)
```
Comment (trivialfis) [inline R-package/R/xgb.DMatrix.R:47]: I will add this back in https://github.com/dmlc/xgboost/pull/9828 .

**Pull request merged**